File: core/src/androidMain/kotlin/ru/mobileup/template/core/map/internal/CurrentLocationMarkerController.kt
package ru.mobileup.template.core.map.internal

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.toArgb
import com.yandex.mapkit.MapKit
import com.yandex.mapkit.MapKitFactory
import com.yandex.mapkit.layers.ObjectEvent
import com.yandex.mapkit.map.IconStyle
import com.yandex.mapkit.map.RotationType
import com.yandex.mapkit.mapview.MapView
import com.yandex.mapkit.user_location.UserLocationLayer
import com.yandex.mapkit.user_location.UserLocationObjectListener
import com.yandex.mapkit.user_location.UserLocationView
import com.yandex.runtime.image.ImageProvider
import ru.mobileup.template.core.R

private val ACCURACY_CIRCLE_COLOR = Color(0x99B6B6B6).toArgb()
private const val DEFAULT_SCALE = 0.2f

internal class CurrentLocationMarkerController(
    private val mapView: MapView
) : UserLocationObjectListener {

    private var mapKit: MapKit = MapKitFactory.getInstance()
    private var userLocationLayer: UserLocationLayer? = mapKit.createUserLocationLayer(mapView.mapWindow)

    init {
        userLocationLayer?.setObjectListener(this)
    }

    fun setVisible(isVisible: Boolean) {
        if (isVisible) {
            mapKit.resetLocationManagerToDefault()
            userLocationLayer?.isVisible = true
        } else {
            userLocationLayer?.isVisible = false
        }
    }

    override fun onObjectAdded(userLocationView: UserLocationView) {
        val context = mapView.context

        userLocationView.accuracyCircle.fillColor = ACCURACY_CIRCLE_COLOR

        userLocationView.arrow.setIcon(
            ImageProvider.fromResource(context, R.drawable.ic_my_location_arrow),
            IconStyle()
                .setRotationType(RotationType.ROTATE)
                .setScale(DEFAULT_SCALE)
        )

        userLocationView.pin.setIcon(
            ImageProvider.fromResource(context, R.drawable.ic_my_location_dot),
            IconStyle()
                .setScale(DEFAULT_SCALE)
        )
    }

    override fun onObjectRemoved(userLocationView: UserLocationView) = Unit

    override fun onObjectUpdated(userLocationView: UserLocationView, event: ObjectEvent) = Unit
}

File: iosApp/iosApp/Source/Map/Subcontrollers/CurrentLocationMarkerController.swift
import YandexMapsMobile
import Shared

final class CurrentLocationMarkerController: NSObject, YMKUserLocationObjectListener {

    // 0x99B6B6B6 (semi-transparent gray), matching the Android accuracy circle color.
    private static let accuracyCircleColor = UIColor(
        red: 0xB6 / 255,
        green: 0xB6 / 255,
        blue: 0xB6 / 255,
        alpha: 0x99 / 255
    )
    private static let defaultScale: NSNumber = 0.2

    private let mapView: YMKMapView
    private let mapKit: YMKMapKit
    private var userLocationLayer: YMKUserLocationLayer?

    init(mapView: YMKMapView) {
        self.mapView = mapView
        self.mapKit = YMKMapKit.sharedInstance()
        super.init()

        let layer = mapKit.createUserLocationLayer(with: mapView.mapWindow)
        layer.setObjectListenerWith(self)
        userLocationLayer = layer
    }

    func setVisible(isVisible: Bool) {
        if isVisible {
            mapKit.resetLocationManagerToDefault()
            userLocationLayer?.setVisibleWithOn(true)
        } else {
            userLocationLayer?.setVisibleWithOn(false)
        }
    }

    func dispose() {
        userLocationLayer?.setVisibleWithOn(false)
        userLocationLayer?.setObjectListenerWith(nil)
        userLocationLayer = nil
    }

    // MARK: - YMKUserLocationObjectListener

    func onObjectAdded(with view: YMKUserLocationView) {
        view.accuracyCircle.fillColor = Self.accuracyCircleColor

        let arrowStyle = YMKIconStyle()
        arrowStyle.rotationType = NSNumber(value: YMKRotationType.rotate.rawValue)
        arrowStyle.scale = Self.defaultScale
        if let arrowImage = UIImage(named: "ic_my_location_arrow") {
            view.arrow.setIconWith(arrowImage, style: arrowStyle)
        }

        let pinStyle = YMKIconStyle()
        pinStyle.scale = Self.defaultScale
        if let pinImage = UIImage(named: "ic_my_location_dot") {
            view.pin.setIconWith(pinImage, style: pinStyle)
        }
    }

    func onObjectRemoved(with view: YMKUserLocationView) {}

    func onObjectUpdated(with view: YMKUserLocationView, event: YMKObjectEvent) {}
}
